### ISSUE TITLE
"E/SQLiteLog: (1) no such table: " occurs when I use "Amplify.DataStore.save()"

### ISSUE THREAD

**Issue by Shumpei0706** [aws-amplify/amplify-android#595]
Hi, I am facing an error again and again when I change my GraphQL schema very slightly from tutorial:

>  "E/SQLiteLog: (1) no such table: "

I only changed the name of the model from `Todo` to `Tod`. I tried to check many cases, but even this small change can cause this error.

Am I supposed to register the schema when it's changed, in addition to running `modelgen`?

**Comment by jamesonwilliams** [comment]
> Do I supposed to register the schema when it's changed in addition to run "modelgen"?

Hi @Shumpei0706 -- yes, if you change the structure/names of the models, you should re-run code generation, and push your changes to the backend, as well.

**Comment by woutermilants** [comment]
I'm having the same issue. 
So first run modelgen and then ? 

My schema is created in AWS, but locally it still fails

**Comment by AntonPolkanov** [comment]
I am getting the same error (_"E/SQLiteLog: (1) no such table"_), but it slightly defferent use case.

I did the getting started tutorial https://docs.amplify.aws/start/q/integration/android and everything worked fine. Then I 
1. added a new type Comment to the gql schema
2. ran modelgen
3. ran the application 
and got "E/SQLiteLog: (1) no such table: Comment" error.

Current schema: 
```
enum Priority {
  LOW
  NORMAL
  HIGH
}

type Comment @model {
  id: ID!
  content: String
}

type Todo @model {
  id: ID!
  name: String!
  priority: Priority
  description: String
}
```
If I delete the Comment type from the schema and do steps (2) and (3) the error goes away.

**Comment by jamesonwilliams** [comment]
The intended behavior of the system is as follows.

__On the development machine, at build time:__ Every set of model schema gets a unique hash. If any of the model schema change, the hash changes.

__On the Android device, at run time:__ If there is no hash in the database, the current model schema hash is stored. If there is a hash in the database, it is compared against the current model schema hash. If they are different, the old database is dropped, and a new one is created, to accommodate the new schema. If the hashes are the same, nothing happens.

It sounds to me like there is a bug, and the system is not correctly re-establishing the database, when the model hash changes.

In theory we have a test for this scenario, [here](https://github.com/aws-amplify/amplify-android/blob/main/aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/ModelUpgradeSQLiteInstrumentedTest.java#L88-L94). It may be too low level to catch the current bug.

**Comment by jamesonwilliams** [comment]
> I have no idea why I should delete my app each time I change my schema. Can someone please explain why is it so?

@AntonPolkanov You shouldn't have to. I think it's a bug.

**Comment by Shumpei0706** [comment]
I am glad to have the same understanding about the issue.

I tried to delete my app and found it worked, I mean my modified schema worked.

I hope this issue resolve.

**Comment by sunithapremjee** [comment]
I changed my schema.graphql table name from Kospetdata to KospetdataUnit. Ran modelgen , rebuilt the app and did amplifypush, Deleted the app on phone and reinstalled app. But still I get "SQLiteLog: (1) no such table: KospetdataUnit". Please direct me if I miss any configuration ASAP.
![Table_Model](https://user-images.githubusercontent.com/9869417/86882474-4e685100-c0be-11ea-9431-aaa6a20de385.png)

**Comment by sunithapremjee** [comment]
Above issue occurs on Android 10 device .Android 8 device gives following exception and app crashes... Can anybody here help me with this issue please.
java.lang.RuntimeException: java.util.concurrent.TimeoutException: The source did not signal an event for 5000 milliseconds and has been terminated.
        at io.reactivex.internal.util.ExceptionHelper.wrapOrThrow(ExceptionHelper.java:46)
        at io.reactivex.internal.observers.BlockingMultiObserver.blockingGetError(BlockingMultiObserver.java:153)
        at io.reactivex.Completable.blockingGet(Completable.java:1302)
        at com.amplifyframework.datastore.AWSDataStorePlugin.initialize(AWSDataStorePlugin.java:7)
        at com.amplifyframework.core.category.Category.initialize(Category.java:15)
        at com.amplifyframework.core.Amplify.lambda$beginInitialization$0(Amplify.java:1)
        at com.amplifyframework.core.-$$Lambda$Amplify$EB8h7rBH19uqUcPr0ZtYS0ww0r4.run(Unknown Source:4)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: java.util.concurrent.TimeoutException: The source did not signal an event for 5000 milliseconds and has been terminated.
        at io.reactivex.internal.observers.BlockingMultiObserver.blockingGetError(BlockingMultiObserver.java:153) 
        at io.reactivex.Completable.blockingGet(Completable.java:1302) 
        at com.amplifyframework.datastore.AWSDataStorePlugin.initialize(AWSDataStorePlugin.java:7) 
        at com.amplifyframework.core.category.Category.initialize(Category.java:15) 
        at com.amplifyframework.core.Amplify.lambda$beginInitialization$0(Amplify.java:1) 
        at com.amplifyframework.core.-$$Lambda$Amplify$EB8h7rBH19uqUcPr0ZtYS0ww0r4.run(Unknown Source:4) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636) 
        at java.lang.Thread.run(Thread.java:764)

**Comment by jamesonwilliams** [comment]
@sunithapremjee Thanks for the additional info. By the way, checkout [GitHub's formatting guide](https://docs.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks).

**Comment by sunithapremjee** [comment]
Android 8 device shows following log in android studio.  Unlike Android 10 device with "SQLiteLog :no such table" error, Android 8 device could create table "KospetdataUnit"

''' Java

2020-07-09 18:19:48.387 30068-30253/com.prodigy.kospet I/KospetAmplifyApplication: Initialized Amplify
2020-07-09 18:19:48.503 30068-30436/com.prodigy.kospet D/AWSMobileClient: Inspecting user state details
2020-07-09 18:19:48.526 30068-30435/com.prodigy.kospet I/amplify:aws-datastore: Creating table: PersistentModelVersion
2020-07-09 18:19:48.528 30068-30435/com.prodigy.kospet I/amplify:aws-datastore: Creating table: PersistentRecord
2020-07-09 18:19:48.529 30068-30435/com.prodigy.kospet I/amplify:aws-datastore: Creating table: LastSyncMetadata
2020-07-09 18:19:48.530 30068-30435/com.prodigy.kospet I/amplify:aws-datastore: Creating table: ModelMetadata
2020-07-09 18:19:48.541 30068-30435/com.prodigy.kospet I/amplify:aws-datastore: Creating table: KospetdataUnit

''''

**Comment by sunithapremjee** [comment]
On Android 8 device I could sign in ,add data to datastore and it successfully synced to cloud first time. But when I take the app next time on calling Amplify.Auth.fetchAuthSession , result.isSignedIn() is true and app crashes due to

'''java
      java.lang.RuntimeException: java.util.concurrent.TimeoutException: The source did not signal an event for 5000 milliseconds 
      and has been terminated.
,,,

On Android 10 device ,first time itself I get 

'''java
        "SQLiteLog :no such table"
'''
Please tell me if I should create a new project and do all configuration one more time.???

**Comment by sunithapremjee** [comment]
@jamesonwilliams following is the autogenerated Model class KospetdataUnit function hashCode() which doesn't include .append("newtablename") which is the only change I made in my Schema (Table name changed from Kospetdata to KospetDataUnit, all other fields remains the same.So the hashCode seems to remain the same in such case. Please correct me if I am wrong.

'''java

 @Override
   public int hashCode() {
    return new StringBuilder()
      .append(getId())
      .append(getTitle())
      .append(getDescription())
      .append(getAudiopath())
      .append(getSensitivity())
      .append(getTimer())
      .toString()
      .hashCode();
  }
'''java

**Comment by jamesonwilliams** [comment]
@sunithapremjee (or anyone else) -- what is the output of this command, before and after you update the model?

```bash
cat app/src/main/java/com/amplifyframework/datastore/generated/model/AmplifyModelProvider.java
```

I'm wondering if it isn't actually changing.

If it _isn't_, try updating the models on the command line with `amplify codegen model`, instead of by using the Gradle plugin.

**Comment by sunithapremjee** [comment]
Android studio shows this warning 
Class 'AmplifyModelProvider' is never used

    On Sunday, July 12, 2020, 04:09:38 PM EDT, Jameson Williams <notifications@github.com> wrote:  
 
 


@sunithapremjee (or anyone else) -- what is the output of this command, before and after you update the model?
cat app/src/main/java/com/amplifyframework/datastore/generated/model/AmplifyModelProvider.java
I'm wondering if it isn't actually changing.

If it isn't, try updating the models on the command line with amplify codegen model, instead of by using the Gradle plugin.

**Comment by sunithapremjee** [comment]
I changed the Model name from KospetdataUnit to KospetUnit.Ran modelgen.

Yes AMPLIFY_MODEL_VERSION in AmplifyModelProvider.java now shows a new distinct value.

**Comment by rjuliano** [comment]
Adding steps to reproduce behavior for reference:
- Given an existing Amplify ModelProvider with `AMPLIFY_MODELS_VERSION=123` that is currently being consumed by an app via the AWSDataStorePlugin.
- Modify the schema and add a new type called `House` for example.
- Regenerate the models. That should generate a new hash for the ModelProvider. `AMPLIFY_MODELS_VERSION=456` for example.
- Reload the app.

The app attempts to initialize the storage adapter using `AMPLIFY_MODELS_VERSION=456`. As part of this process, [when calling getInsertSqlPreparedStatements](https://github.com/aws-amplify/amplify-android/blob/4ad707a4c8f0975d271e3631deb4d9de6f3ef000/aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapter.java#L229), we get an exception because the underlying db doesn't have the table for the new type (evidently, the table has to be in the database for us to be able to prepare these statements). Because of the exception, we never get to the point [where we do the model updates](https://github.com/aws-amplify/amplify-android/blob/4ad707a4c8f0975d271e3631deb4d9de6f3ef000/aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapter.java#L236). 

I have created a [branch](https://github.com/aws-amplify/amplify-android/tree/rjuliano/schema_upgrade_fix) where I updated the [unit test](https://github.com/aws-amplify/amplify-android/commit/959dc270006404f95a0a6f0d9ab6629def1d2abb) to reflect this use case. The previous version of the test was just simulating a change in hash, but the schemas themselves were the same, so it didn't catch this use case.

Kind of a chicken and the egg thing. We have to read the model version from the database, to see if we need to update the schema. In order to do that, we need initialize the storage adapter. To initialize the storage adapter, you need the updated schema in place. Maybe we need to separate "System" tables from application tables.

**Comment by raphkim** [comment]
> The app attempts to initialize the storage adapter using AMPLIFY_MODELS_VERSION=456. As part of this process, when calling getInsertSqlPreparedStatements, we get an exception because the underlying db doesn't have the table for the new type (evidently, the table has to be in the database for us to be able to prepare these statements). Because of the exception, we never get to the point where we do the model updates.

#718 addresses the above concern by simply removing the statement caching.

Ever since this little optimization logic was added, we were skeptical about how much performance boost we are actually getting out of it. However, now that it got to a point where it is introducing new bugs, it became pretty clear that the minute optimization simply isn't worth the effort.

**Comment by rjuliano** [comment]
Included in version 1.1.3
